vpc/allocate/ip_pool_manager_test.go

```diff
@@ -0,0 +1,150 @@
+package allocate
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"reflect"
+	"runtime"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/Netflix/titus-executor/fslocker"
+	vpcContext "github.com/Netflix/titus-executor/vpc/context"
+	"github.com/Netflix/titus-executor/vpc/ec2wrapper"
+	"github.com/aws/aws-sdk-go/aws/client"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const fakeMac = "fake-mac"
+
+var (
+	_ ec2wrapper.NetworkInterface = (*testNetworkInterface)(nil)
+)
+
+type testNetworkInterface struct {
+	ipv4Addresses []string
+	freeIPs       func([]string) error
+	refresh       func() error
+}
+
+func (tni *testNetworkInterface) FreeIPv4Addresses(ctx context.Context, _ client.ConfigProvider, deallocationList []string) error {
+	if tni.freeIPs != nil {
+		return tni.freeIPs(deallocationList)
+	}
+	panic("Free IPs not implemented")
+}
+
+func (*testNetworkInterface) GetDeviceNumber() int {
+	panic("implement me")
+}
+
+func (*testNetworkInterface) GetInterfaceID() string {
+	panic("implement me")
+}
+
+func (*testNetworkInterface) GetSubnetID() string {
+	panic("implement me")
+}
+
+func (*testNetworkInterface) GetMAC() string {
+	return fakeMac
+}
+
+func (*testNetworkInterface) GetSecurityGroupIds() map[string]struct{} {
+	panic("implement me")
+}
+
+func (tni *testNetworkInterface) GetIPv4Addresses() []string {
+	return tni.ipv4Addresses
+}
+
+func (tni *testNetworkInterface) Refresh() error {
+	if tni.refresh != nil {
+		return tni.refresh()
+	}
+	panic("Refresh not implemented")
+}
+
+func testFreeIPsOneIP(t *testing.T, ctx *vpcContext.VPCContext, tni *testNetworkInterface, ipPoolManager *IPPoolManager) {
+	tni.ipv4Addresses = []string{"1.2.3.4"}
+	assert.NoError(t, ipPoolManager.DoGc(ctx, time.Second))
+	assert.NoError(t, ipPoolManager.DoGc(ctx, time.Second))
+	time.Sleep(3 * time.Second)
+	assert.NoError(t, ipPoolManager.DoGc(ctx, time.Second))
+}
+
+func testFreeIPsTwoIP(t *testing.T, ctx *vpcContext.VPCContext, tni *testNetworkInterface, ipPoolManager *IPPoolManager) {
+	ipPoolManager.ipRefreshSleepInterval = 0
+	var ipList = []string{"1.2.3.4", "5.6.7.8", "9.8.10.11"}
+	called := false
+	tries := 3
+	// Force 3 tries before IPs go away
+	tni.refresh = func() error {
+		if tries > 0 {
+			tni.ipv4Addresses = ipList[:1]
+		}
+		tries--
+		return nil
+	}
+	tni.freeIPs = func(deallocationList []string) error {
+		called = true
+		assert.Equal(t, ipList[1:], deallocationList)
+		return nil
+	}
+	tni.ipv4Addresses = ipList
+	assert.NoError(t, ipPoolManager.DoGc(ctx, time.Second))
+	assert.NoError(t, ipPoolManager.DoGc(ctx, time.Second))
+	time.Sleep(3 * time.Second)
```

Comment:
I know you are just starting on this, but adding an abstraction for an injectable `type Clock interface` (or `func Now() time.Time`) that the `IPPoolManager` uses will help keep execution times for these tests manageable. Something to keep an eye on as you add more tests.

Reply:
that way you can control the wallclock in tests with things like `clock.AdvanceBy(3 * time.Second)` instead of `time.Sleep(3 * time.Second)`.

Reply:
Yeah, there's a bunch of places in the tests where I think it might make a lot of sense to mock-out the clock. In standalone tests too. 

Is there a clock library you know that allows that, but with a way to avoid skipping over multiple events?

Reply:
Unfortunately not, I haven't used any in Go.